internal/allocator/allocator.go

```diff
@@ -263,69 +479,79 @@ func (a *Allocator) AllocateFromPool(svcKey string, svc *v1.Service, serviceIPFa
 		return alloc.ips, nil
 	}
 
-	pool := a.pools.ByName[poolName]
-	if pool == nil {
-		return nil, fmt.Errorf("unknown pool %q", poolName)
+	serviceIPFamilyPolicy := v1.IPFamilyPolicySingleStack
+	if svc.Spec.IPFamilyPolicy != nil {
+		serviceIPFamilyPolicy = *(svc.Spec.IPFamilyPolicy)
 	}
 
-	ips := []net.IP{}
-	ipfamilySel := make(map[ipfamily.Family]bool)
+	return a.AllocateFromPoolForFamilyAndPolicy(svcKey, svc, serviceIPFamily, serviceIPFamilyPolicy, poolName, ports, sharingKey, backendKey)
+}
 
-	switch serviceIPFamily {
-	case ipfamily.DualStack:
-		ipfamilySel[ipfamily.IPv4], ipfamilySel[ipfamily.IPv6] = true, true
-	default:
-		ipfamilySel[serviceIPFamily] = true
+// AllocateFromPoolForFamilyAndPolicy assigns an available IP from pool to service, with a specified serviceIPFamily
+// and suitable for the service IPFamilyPolicy.
+func (a *Allocator) AllocateFromPoolForFamilyAndPolicy(
```

Comment:
there are a few too many Allocate* functions. Let's squash this with the one above as it always ends up in here and this is short enough.